### PR TITLE
788 copy route path part tool

### LINKED ISSUE #788: Copy routePath part tool: adds an ability to copy part from an existing routePath
BE PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/51
FE PR: https://github.com/HSLdevcom/jore-map-ui/pull/830

Make a completely new tool "copyRoutePathPartTool" next to extend routePath tool.

Phase 1 - select start node
* when actived, routePath's end nodes become highlighted
* user has to select highlighted routePath node

Phase 2 - select end node
* selected routePath node becomes highlighted (with green for example)
* if user clicks network's node, all routePaths from which user can copy are listed at sidebar
   * if no routePath is found, empty results text is shown at sidebar

Use can click copy button that copies the routePath seqments or open routePath at new window button

### PR DESCRIPTION
**Pull request by Divinare**
Closes #788

BE PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/51

Go through commit messages to see what was done. Firsly I implemented copy routePath seqment view in a popup, then reverted it to be shown in sidebar
* renamed StartMarker as Marker + renamed mapIcons folder as markers folder caused some additional file changes
* made a new card #832 to prevent user interaction while copying of seqments is in progress.

### CODE REVIEW

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 1]
```diff
@@ -221,17 +229,27 @@ class UpsertRoutePathLayer extends Component<IRoutePathLayerProps, IRoutePathLay
...
+        return(
             <>
                 {this.renderRoutePathLinks()}
                 {this.renderLinkDecorator()}
-                {this.renderStartMarker()}
+                {this.renderMarker()}
                 { neighborLinks &&
                     <RoutePathNeighborLinkLayer />
                 }
+                { isCopySeqmentLayerVisible &&
```
Comment (Divinare) [inline src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: isRoutePathCopySeqmentLayerVisible

File: src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                    <div
+                        className={s.icon}
+                        title={`Kopioi reitin ${routePath.routeId} reitin suunnan segmentti`}
+                        onClick={this.copySeqments(routePath)}
+                    >
+                        <FiCopy />
+                    </div>
+                    <div
+                        className={s.icon}
+                        title={`Avaa reitin ${routePath.routeId} reitin  suunta uuteen ikkunaan`}
```
Comment (Divinare) [inline src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx]: Two spaces: reitin_ _suunta

File: src/stores/routePathCopySeqmentStore.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @action
+    public clear = () => {
+        this._endNode = null;
+        this._startNode = null;
+        this._routePaths = [];
+    }
+
+    public getLinksToCopy = (
```
Comment (Divinare) [inline src/stores/routePathCopySeqmentStore.ts]: getSeqmentsToCopy

File: src/stores/routePathCopySeqmentStore.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        const startLinkOrderNumber = this._getStartLinkOrderNumber(routePath.links, startNodeId);
+        const endLinkOrderNumber = this._getEndLinkOrderNumber(routePath.links, endNodeId);
+
+        return routePath.links.filter((link: ICopySeqmentLink) => {
+            return link.orderNumber >= startLinkOrderNumber
+                && link.orderNumber <= endLinkOrderNumber;
+        });
+    }
+
+    public getLinksNotToCopy = (
```
Comment (Divinare) [inline src/stores/routePathCopySeqmentStore.ts]: getSeqmentsNotToCopy

File: src/stores/routePathCopySeqmentStore.ts (hunk 3 above)
Comment (Divinare) [reply src/stores/routePathCopySeqmentStore.ts]: or getSeqmentLinksToCopy?

**Comment by Divinare** [comment]
Added a few review comments for myself. This task can be reviewed still, I can fix those at the same time I fix other review comments

**Comment by kronken** [comment]
Founds some bugs/ things to fix:
- App randomly throws error when searching for paths between two points, tried to understand when it comes, but I haven't understood any patterns. Should be easy to reproduce, for me it comes 30% of the time.
- Add tool hint
- Unable to search path that begins from route path beginning
- Unable to search path between two highlighted nodes

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 5]
```diff
@@ -6,21 +6,28 @@ import IRoutePathLink from '~/models/IRoutePathLink';
 import { createCoherentLinesFromPolylines } from '~/util/geomHelper';
 import INode from '~/models/INode';
 import { RoutePathStore, RoutePathViewTab } from '~/stores/routePathStore';
+import { RoutePathCopySeqmentStore } from '~/stores/routePathCopySeqmentStore';
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: Changed this now everywhere in code

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 6]
```diff
@@ -85,27 +92,28 @@ class UpsertRoutePathLayer extends Component<IRoutePathLayerProps, IRoutePathLay
...
             // Allow click event for highlighted nodes only
             if (isNodeHighlighted) {
-                const clickParams: IExtendRoutePathNodeClickParams = { node, linkOrderNumber };
+                const clickParams: IEditRoutePathLayerNodeClickParams = { node, linkOrderNumber };
                 onNodeClick = () =>
                     EventManager.trigger('nodeClick', clickParams);
             }
         } else {
             // Prevent default click if there are neighbors on map
-            if (this.props.routePathStore!.neighborLinks.length === 0) {
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: Yes it wasn't needed. Good you noticed I left a comment behind

File: src/util/leafletUtils.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import ReactDOMServer from 'react-dom/server';
+import * as L from 'leaflet';
+import * as s from './leafletUtils.scss';
+
+const DEFAULT_POPUP_OFFSET = -30;
+
+const createDivIcon = (html: any, className?: any, popupOffset?: number) => {
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/util/leafletUtils.ts]: Its now a lot better thanks

File: src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    routePathCopySeqmentStore?: RoutePathCopySeqmentStore;
+    toolbarStore?: ToolbarStore;
+}
+
+@inject('routePathStore', 'routePathCopySeqmentStore', 'toolbarStore')
+@observer
+class RoutePathCopySeqmentView extends React.Component<IRoutePathCopySeqmentViewProps> {
+
+    private renderRoutePathRow = (routePath: ICopySeqmentRoutePath, key: string) => {
+        return (
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx]: I don't see this as a problem when extra return = (). Yes, it is shorter to write but not that big deal. Also if in the future I need to put something there, I might have to revert it back to use return = ()

File: src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                </div>
+            </div>
+        );
+    }
+
+    private setHighlightedRoutePath = (routePath: ICopySeqmentRoutePath|null) => () => {
+        this.props.routePathCopySeqmentStore!.setHighlightedRoutePath(routePath);
+    }
+
+    private renderTextRow = (routePath: ICopySeqmentRoutePath) => {
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx]: See upper comment

File: src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    private copySeqment = async (routePathLinkId: number) => {
+        const routePathLink: IRoutePathLink = await RoutePathLinkService
+            .fetchRoutePathLink(routePathLinkId);
+        this.props.routePathStore!.addLink(routePathLink);
+    }
+
+    private openRoutePathInNewTab = (routePath: ICopySeqmentRoutePath) => () => {
+        const routePathLink = routeBuilder
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/sidebar/routePathView/RoutePathCopySeqmentView.tsx]: This is how 90% of our links created by routeBuilder are named. If I change this, I should also change it everywhere else... But yes I also agree that url is better than link

**Comment by Divinare** [comment]
Review fixes done. This can be reviewed again.

- App randomly throws error when searching for paths between two points, tried to understand when it comes, but I haven't understood any patterns. Should be easy to reproduce, for me it comes 30% of the time.
   - this was a nasty bug, fixed with reseting highlightedRoutePath when store is cleared
- Add tool hint
   - there is now toolHelp
- Unable to search path that begins from route path beginning
   - marker gets "isClickable" as props
- Unable to search path between two highlighted nodes
   - there are two buttons to set start / end marker

**Comment by Divinare** [comment]
Fixes done:
- color coding of markers
- unclickable map bug
- bug: copying of 1 segment not found
- displaying error messages when nodes are incorrectly placed

This can be reviewed again.

File: src/components/sidebar/routePathView/RoutePathCopySegmentView.tsx [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                orderNumber += 1;
+            }
+        } else if (isOppositeDirection) {
+            const orderNumber = this.props.routePathStore!.routePath!.routePathLinks!
+                .find(link => link.startNode.id === copyEndNodeId)!.orderNumber;
+            for (let i = segmentsToCopy.length - 1; i >= 0; i -= 1) {
+                await this.copySegment(segmentsToCopy[i].routePathLinkId, orderNumber);
+            }
+        } else {
+            // sanity check
```
Comment (kronken) [inline src/components/sidebar/routePathView/RoutePathCopySegmentView.tsx]: // Should not occur

File: src/components/sidebar/routePathView/RoutePathCopySegmentView.tsx [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+                { routePaths.length === 0 ? (
+                    <div className={s.messageContainer}>
+                        Kopioitavia reitinsuunnan segmenttejä ei löytynyt valitulta väliltä.
+                        Kokeile asettaa pienempi kopioitava väli.
+                    </div>
+                ) : (
+                    <>
+                    { routePaths.map((routePath, index) => {
+                        return this.renderRoutePathRow(
+                            routePath, `row-${index}-${routePath.routeId}'`);
```
Comment (kronken) [inline src/components/sidebar/routePathView/RoutePathCopySegmentView.tsx]: The string ends with a '
Suppose it is a mistake?

Comment (Divinare) [reply src/components/sidebar/routePathView/RoutePathCopySegmentView.tsx]: Mistake yes, but I guess it didn't cause any bugs. Fixed it tho.

**Comment by Divinare** [comment]
Review fixes done.

**Comment by Divinare** [comment]
Found orderNumber bug & fixed it

**Pull request merged**